src/window_draw_list.rs

```diff
@@ -0,0 +1,617 @@
+            pub fn add_triangle<P1, P2, P3, C>(
+                &self,
+                p1: P1,
+                p2: P2,
+                p3: P3,
+                c: C,
+            ) -> Triangle<'ui, $T>
+            where
+                P1: Into<ImVec2>,
+                P2: Into<ImVec2>,
+                P3: Into<ImVec2>,
+                C: Into<ImColor>,
+            {
+                Triangle::new(self, p1, p2, p3, c)
+            }
+
+            /// Returns a circle with the given `center`, `radius` and `color`.
+            pub fn add_circle<P, C>(&self, center: P, radius: f32, color: C) -> Circle<'ui, $T>
+            where
+                P: Into<ImVec2>,
+                C: Into<ImColor>,
+            {
+                Circle::new(self, center, radius, color)
+            }
+
+            /// Returns a Bezier curve stretching from `pos0` to `pos1`, whose
+            /// curvature is defined by `cp0` and `cp1`.
+            pub fn add_bezier_curve<P1, P2, P3, P4, C>(
+                &self,
+                pos0: P1,
+                cp0: P2,
+                cp1: P3,
+                pos1: P4,
+                color: C,
+            ) -> BezierCurve<'ui, $T>
+            where
+                P1: Into<ImVec2>,
+                P2: Into<ImVec2>,
+                P3: Into<ImVec2>,
+                P4: Into<ImVec2>,
+                C: Into<ImColor>,
+            {
+                BezierCurve::new(self, pos0, cp0, cp1, pos1, color)
+            }
+
+            /// Push a clipping rectangle on the stack, run `f` and pop it.
+            ///
+            /// Clip all drawings done within the closure `f` in the given
+            /// rectangle.
+            pub fn with_clip_rect<P1, P2, F>(&self, min: P1, max: P2, f: F)
+            where
+                P1: Into<ImVec2>,
+                P2: Into<ImVec2>,
+                F: FnOnce(),
+            {
+                unsafe {
+                    sys::ImDrawList_PushClipRect(self.draw_list(), min.into(), max.into(), false)
+                }
+                f();
+                unsafe { sys::ImDrawList_PopClipRect(self.draw_list()) }
+            }
+
+            /// Push a clipping rectangle on the stack, run `f` and pop it.
+            ///
+            /// Clip all drawings done within the closure `f` in the given
+            /// rectangle. Intersect with all clipping rectangle previously on
+            /// the stack.
+            pub fn with_clip_rect_intersect<P1, P2, F>(&self, min: P1, max: P2, f: F)
+            where
+                P1: Into<ImVec2>,
+                P2: Into<ImVec2>,
+                F: FnOnce(),
+            {
+                unsafe {
+                    sys::ImDrawList_PushClipRect(self.draw_list(), min.into(), max.into(), true)
+                }
+                f();
+                unsafe { sys::ImDrawList_PopClipRect(self.draw_list()) }
+            }
+        }
+    };
+}
+
+impl_draw_list_methods!(WindowDrawList);
```

Comment:
I actually thought about doing that, so let me explain the reasoning behind my choice.

If the methods are defined on the trait, then the user of the lib will be required to import the `DrawAPI` trait to be able to use any of the methods. As adding a supplementary import is not convenient I chose to duplicate all method definitions with macros. With this design,`DrawAPI` is an internal trait used only internally for its `draw_list` method.

What do you think would be the best for the user's interest?

Reply:
Yeah, I understand this reasoning and I've seen this pattern in some libs before (including imgui-rs itself ;) ). However, using macros to copy-paste code means that we're also ignoring potentially beneficial abstraction.

In this case, imagine that you want to write a function that draws something:

```
fn draw_my_amazing_thing<'ui>(dl: &WindowDrawList<'ui>) {
  dl.add_line(...);
  // ...
}
```

This is all fine and you might be happy with this function, but now we've tied this function to WindowDrawList, so it won't work with ChannelsSplit even though it too has the `add_line` method. This seems strange if WindowDrawList and ChannelsSplit are supposed to have the same functionality.

What if you write a new trait to support both types? That seems very strange since we already have DrawAPI which could be used to abstract the common functionality.

What if you require that `dl` implements `DrawAPI` and use the `Line::new` function in the implementation instead of `add_line`? That works because it supports anything that implements DrawAPI. But now what's the point of having `add_line`?

In my opinion using macros to copy-paste code only makes sense when you want to copy some boilerplate into types that are so *different* that you know there's no need to abstract over them.

So, I'd prefer just having the functions in DrawAPI to avoid all these issues. Yes, the user has to import the trait, but this is just how things work in Rust. Is it a problem that you need to import `std::io::Read` to read from a `std::fs::File`?

However, don't start changing things just yet, because I think we have a slightly different approach and abolish DrawAPI completely. I'll explain this in another comment :smiley:

Reply:
That makes so much sense. Thank you very much for the explanation :) . I feel a better Rustacean now. Looking forward to your idea about a different approach then!